### PR TITLE
improvement: ZENKO-2535 add microVersionId to ObjectMD

### PR DESCRIPTION
**Pull request by jonathan-gramain**
Add a new microVersionId field that is a hex-encoded field of 64
bits randomly generated.

It can be useful to:

- force updates in MongoDB when no other metadata changes, by updating
  the microVersionId field

- manage concurrent updates, by adding a condition on the
  microVersionId and updating it for each metadata update to perform

In order for the change to be less intrusive, it is an optional field:
if ObjectMD.updateMicroVersionId() is not called by the client, the
metadata will not contain a microVersionId field. Clients will call
this function when needed on a case-by-case basis.

### CODE REVIEW

File: lib/models/ObjectMD.js [hunk 1]
```diff
@@ -1028,6 +1030,33 @@ class ObjectMD {
...
+     * no other metadata field changes) or to manage conflicts during
+     * concurrent updates, using conditions on the microVersionId
+     * field.
+     *
+     * It's a base64-encoded field of 64 bits randomly generated
+     *
+     * @return {ObjectMD} itself
+     */
+    updateMicroVersionId() {
+        this._data.microVersionId = crypto.randomBytes(8).toString('base64');
```
Comment (vrancurel) [inline lib/models/ObjectMD.js]: I would prefer 'hex' than 'base64' because it is more readable (although a bit longer).

**Review by vrancurel — CHANGES_REQUESTED** [review]

File: lib/models/ObjectMD.js (hunk 1 above)
Comment (jonathan-gramain) [reply lib/models/ObjectMD.js]: not convinced, but ok, I don't mind changing it.

Comment (jonathan-gramain) [reply lib/models/ObjectMD.js]: done

File: tests/unit/models/ObjectMD.js [hunk 2]
```diff
@@ -257,6 +257,23 @@ describe('ObjectMD class setters/getters', () => {
...
         assert.strictEqual(md.getUserMetadata(), undefined);
     });
+
+    it('ObjectMD::microVersionId', () => {
+        assert.strictEqual(md.getMicroVersionId(), null);
+
+        const generatedIds = {};
+        for (let i = 0; i < 100; ++i) {
+            md.updateMicroVersionId();
+            generatedIds[md.getMicroVersionId()] = 1;
```
Comment (miniscruff) [inline tests/unit/models/ObjectMD.js]: If all you care about is the key here you can use a Set instead of an Object

File: tests/unit/models/ObjectMD.js [hunk 3]
```diff
@@ -257,6 +257,23 @@ describe('ObjectMD class setters/getters', () => {
...
+        }
+        // all generated IDs should be different
+        assert.strictEqual(Object.keys(generatedIds).length, 100);
+        Object.keys(generatedIds).forEach(key => {
+            // length is always 16 in hex because leading 0s are
+            // also encoded in the 8-byte random buffer.
+            assert.strictEqual(key.length, 16);
```
Comment (miniscruff) [inline tests/unit/models/ObjectMD.js:276]: Instead of comparing lengths, you could mock the crypto call to ensure it is calling it with specific params

File: tests/unit/models/ObjectMD.js [hunk 4]
```diff
@@ -257,6 +257,23 @@ describe('ObjectMD class setters/getters', () => {
...
+    it('ObjectMD::microVersionId', () => {
+        assert.strictEqual(md.getMicroVersionId(), null);
+
+        const generatedIds = {};
+        for (let i = 0; i < 100; ++i) {
+            md.updateMicroVersionId();
+            generatedIds[md.getMicroVersionId()] = 1;
+        }
+        // all generated IDs should be different
+        assert.strictEqual(Object.keys(generatedIds).length, 100);
```
Comment (miniscruff) [inline tests/unit/models/ObjectMD.js]: Although unlikely this does have a chance to fail due to getting the same random 8 bytes out of 100, I am sure the odds are incredibly tiny but I would also argue that this doesn't really validate our code much either. Simply mocking the random call will have the same affect here.

File: tests/unit/models/ObjectMD.js [hunk 5]
```diff
@@ -257,6 +257,23 @@ describe('ObjectMD class setters/getters', () => {
         md.clearMetadataValues();
         assert.strictEqual(md.getUserMetadata(), undefined);
     });
+
+    it('ObjectMD::microVersionId', () => {
+        assert.strictEqual(md.getMicroVersionId(), null);
```
Comment (miniscruff) [inline tests/unit/models/ObjectMD.js:262]: This line looks like a separate test, that it defaults to none and that you can update it.

File: lib/models/ObjectMD.js [hunk 6]
```diff
@@ -1028,6 +1030,33 @@ class ObjectMD {
...
+        return this;
+    }
+
+    /**
+     * Get the microVersionId field, or null if not set
+     *
+     * @return {string|null} the microVersionId field if exists, or
+     * {null} if it does not exist
+     */
+    getMicroVersionId() {
```
Comment (miniscruff) [inline lib/models/ObjectMD.js:1063]: Way outside the scope of this change but javascript now has get/set methods for classes https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/get and https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/set that would do wonders here.

File: tests/unit/models/ObjectMD.js (hunk 4 above)
Comment (jonathan-gramain) [reply tests/unit/models/ObjectMD.js]: The idea of this test is to ensure that we get a different ID every time it's called, it's more of a smoke test than anything else.

I'm not convinced mocking the random call would add much value, since there is also value in running the actual random code that is being used in production, and it's possible to call it within a unit test.

File: tests/unit/models/ObjectMD.js (hunk 3 above)
Comment (jonathan-gramain) [reply tests/unit/models/ObjectMD.js:276]: see answer below

File: tests/unit/models/ObjectMD.js (hunk 4 above)
Comment (miniscruff) [reply tests/unit/models/ObjectMD.js]: Basically by mocking an existing method we piggyback on all the testing and validation of that method ( or anything really but here its a method ). `crypto.randomBytes(8)` is going to be a very well tested method, simply by mocking and asserting we called that method is plenty of value to me.

Creating a smoke test, or length check or any repeated test here is going to be much less exhaustive when compared to what the crypto package already has to go through... probably. So I would argue being lazy here has the most value.

tl;dr having a test that says we call the crypto package is pretty convincing to me its well tested.

File: lib/models/ObjectMD.js (hunk 6 above)
Comment (miniscruff) [reply lib/models/ObjectMD.js:1063]: Going to resolve this now since there isnt really any action to take here, but will keep it in mind if any refactoring time is allocated.

Comment (miniscruff) [reply lib/models/ObjectMD.js:1063]: @rahulreddy fyi

**Review by vrancurel — APPROVED** [review]

File: lib/models/ObjectMD.js [hunk 7]
```diff
@@ -1028,6 +1030,33 @@ class ObjectMD {
         return this;
     }
 
+    /**
+     * Create or update the microVersionId field
+     *
+     * This field can be used to force an update in MongoDB (in case
+     * no other metadata field changes) or to manage conflicts during
+     * concurrent updates, using conditions on the microVersionId
+     * field.
```
Comment (vrancurel) [inline lib/models/ObjectMD.js]: Would be interesting to mention that it is also useful to detect a change when fields change but object version does not change (e.g. when ingesting a putObjectTagging coming from S3C to Zenko).

Comment (vrancurel) [reply lib/models/ObjectMD.js]: ... And bring clean active-active semantics for S3 calls that are not clean (like putObjectTagging for instance).

File: tests/unit/models/ObjectMD.js [hunk 8]
```diff
@@ -257,6 +257,25 @@ describe('ObjectMD class setters/getters', () => {
...
+
+    it('ObjectMD::microVersionId set', () => {
+        const generatedIds = new Set();
+        for (let i = 0; i < 100; ++i) {
+            md.updateMicroVersionId();
+            generatedIds.add(md.getMicroVersionId());
+        }
+        // all generated IDs should be different
+        assert.strictEqual(generatedIds.size, 100);
+        generatedIds.forEach(key => {
```
Comment (naren-rajendran) [inline tests/unit/models/ObjectMD.js:273]: Could checking length of `microVersionId` be a separate test? It will be cleaner!

Comment (jonathan-gramain) [reply tests/unit/models/ObjectMD.js:273]: IMHO this is a good place to check that the length of all entries is good as well as we just generated the set of version IDs. I don't think it deserves a test on its own.

**Review by naren-rajendran — APPROVED** [review]

File: lib/models/ObjectMD.js (hunk 7 above)
Comment (jonathan-gramain) [reply lib/models/ObjectMD.js]: done

**Comment by jonathan-gramain** [comment]
@bert-e approve

**Review by rahulreddy — APPROVED** [review]
